### PR TITLE
Optimized shooting method

### PR DESCRIPTION
**Pull request by JonasKoziorek**
I implemented recent paper that searches for UPOs via optimization [(link)](https://arxiv.org/abs/1405.5347). The paper also includes python implementation in the appendix. 

My implementation has a few potential problems, that should be addressed:

* I want to avoid creating `ds` in every iteration https://github.com/JuliaDynamics/PeriodicOrbits.jl/blob/507234f624f09fd77be675f4baed4541bd6d3dff/src/algorithms/optimized_shooting.jl#L94-L95 This should be possible by simply not integrating from 0 to 1 but from 0 to T. However I couldn't make it to work for some reason.
* I would like to use other solvers than `RKO65` https://github.com/JuliaDynamics/PeriodicOrbits.jl/blob/507234f624f09fd77be675f4baed4541bd6d3dff/src/algorithms/optimized_shooting.jl#L50-L51 but all other solvers I tried simply weren't working. In the paper they use simple explicit 5-th order runge kutta. I think that choice of solver might be the cause of problems in Saiki too.
* I know that `DynamicalSystemsBase.jl` has `OrdinaryDiffEq.jl` as a dependency but the docs [here](https://juliadynamics.github.io/DynamicalSystemsBase.jl/dev/#CoupledODEs) say that to access the solvers, it has to be imported by the user. I wanted to change the default solver so I added the `OrdinaryDiffEq.jl` as a dependency again. I don't know if this is optimal. Is there a way to access the various solvers internally through `DynamicalSystemsBase.jl`?

### CODE REVIEW

**Comment by Datseris** [comment]
cc also @rveltz : if you know this algorithm, let's make sure that we aren't duplicating effort. If something like this is in BFKit, or parts of the code have been tried in BFKit, let's reuse the knowledge.

File: src/algorithms/optimized_shooting.jl [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+
+"""
+    OptimizedShooting(; kwargs...)
+
+A shooting method [Dednam2014](@cite) combined with Levenberg-Marquardt optimization 
+to find periodic orbits of continuous systems.
+
+## Keyword arguments
+- `Δt::Float64=1e-6`: explicit ODE solver time step. This should correspond to the time step used in the ODE solver specified in the `CoupledODEs` object.
```
Comment (Datseris) [inline src/algorithms/optimized_shooting.jl:13]: This is confusing to me. Is this the number you use when you do `step!(ds, Δt)`? If so, this number is unrelated with the solver step.

File: src/algorithms/optimized_shooting.jl [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+- `p::Int64=2`: `p*dimension(ds)` is the number of points in the residual `R`.
```
Comment (Datseris) [inline src/algorithms/optimized_shooting.jl:14]: Don't use `p` here. Use something else. `p` is a reserved symbol throghout DynamicalSystems.jl used to represent system parameters.

File: docs/make.jl [hunk 3]
```diff
@@ -4,6 +4,8 @@ using PeriodicOrbits
 
 pages = [
     "index.md",
+    "algorithms.md",
```
Comment (Datseris) [inline docs/make.jl:7]: It's good that you added docs, but in the future we should keep PRs separate: each PR is one particular feature/algorithm/fix.

File: docs/refs.bib [hunk 4]
```diff
@@ -0,0 +1,150 @@
+@string{aamop = "Adv. At. Mol. Opt. Phys."}
```
Comment (Datseris) [inline docs/refs.bib:1]: Delete all of this `@string` stuff. Only the articles should be listed here.

File: src/PeriodicOrbits.jl [hunk 5]
```diff
@@ -1,10 +1,18 @@
 module PeriodicOrbits
 
+# Use the README as the module docs
+@doc let
+    path = joinpath(dirname(@__DIR__), "README.md")
+    include_dependency(path)
+    read(path, String)
+end PeriodicOrbits
+
```
Comment (Datseris) [inline src/PeriodicOrbits.jl:9]: This is actually not necessary, I just learned. Julia does this automatically for you.

Comment (Datseris) [reply src/PeriodicOrbits.jl:9]: So you can delete this code block.

File: src/algorithms/optimized_shooting.jl [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+```
+alg = OptimizedShooting(Δt=1/(2^6), p=3)
+ds = CoupledODEs(dynamic_rule, state, params; diffeq = (alg=RKO65(), dt=alg.Δt))
+```
+
+Using other ODE solvers may lead to divergence.
+
+"""
+@kwdef struct OptimizedShooting <: PeriodicOrbitFinder
+    Δt ::Float64 = 1e-6
```
Comment (Datseris) [inline src/algorithms/optimized_shooting.jl:58]: Try to have consistent formatting. Don't have spaces between `::`. I will try to set up an automatic formatter for JuliaDynamics in the meantime.

**Review by Datseris — COMMENTED** [review]
Be sure you merge `main` branch here to update the commits.

File: docs/make.jl (hunk 3 above)
Comment (Datseris) [reply docs/make.jl:7]: We haven't added docs yet to PeriodicOrbits.jl so we could have started this in another PR.

Comment (Datseris) [reply docs/make.jl:7]: Having stuff separated makes PRs easier to review.

File: docs/make.jl [hunk 7]
```diff
@@ -13,4 +15,11 @@ Downloads.download(
...
-build_docs_with_style(pages, PeriodicOrbits)
\ No newline at end of file
+using DocumenterCitations
+
+bib = CitationBibliography(
+    joinpath(@__DIR__, "refs.bib");
+    style=:authoryear
+)
+
+build_docs_with_style(pages, PeriodicOrbits;bib)
```
Comment (Datseris) [inline docs/make.jl:25]:
```suggestion
build_docs_with_style(pages, PeriodicOrbits; bib)
```

formatting again. I will set up auto formatter.

File: src/PeriodicOrbits.jl (hunk 5 above)
Comment (JonasKoziorek) [reply src/PeriodicOrbits.jl:9]: Deleting results in a build error

```
┌ Error: no docs found for 'PeriodicOrbits' in `@docs` block in src/index.md:1-3
│ ```@docs
│ PeriodicOrbits
│ ```
└ @ Documenter ~/.julia/packages/Documenter/qoyeC/src/utilities/utilities.jl:44
```

Can you link some example package where it works?

File: src/algorithms/optimized_shooting.jl (hunk 1 above)
Comment (JonasKoziorek) [reply src/algorithms/optimized_shooting.jl]: In the paper `Δt` is both the integration step and also the time step used when creating the residual. I have two ideas on how to make this better:
1. say that `Δt` is the time step used to create the residual. Notify user that the original paper used the same value as integration step in their explicit ODE solver
2. remove the `Δt` keyword. Notify user that he should use an explicit solver and specify `dt`. This `dt` is then used when creating the residual instead of `Δt`.

**Comment by Datseris** [comment]
@JonasKoziorek can you please write a brief summary of the status of this PR? Does the algorithm work, does it not? If not, when does it not work?

**Comment by JonasKoziorek** [comment]
> @JonasKoziorek can you please write a brief summary of the status of this PR? Does the algorithm work, does it not? If not, when does it not work?

The algorithm works. I have found several UPOs of Lorenz and Roessler systems. One example can be seen in the docs. 
These are the issues before finishing the PR: 
1. find out which ODE solver can be used - right now I only had success with RKO65
2. don't create a new dynamical systems in every step of the optimization process
3. allow the user to modify the residual:
	The residual implicitly depends on these quantities: point in a state space `x,y,z...`, period `T` and parameters of the ds `p`.
	This residual can be changed to achieve different goals.
	One form of the residual is looking for optimal values of `x,y,z...` and `T` with `p` fixed. (we have this right now)
	Other form of the residual would be to look for optimal `p` but leave `x,y,z...` and `T` fixed. 
	This would allow the user to find the parameters `p` which have a PO of period `T` going through point `x,y,z...`.
	In principle any of the quantities can be fixed and the others left for optimization.

**Comment by JonasKoziorek** [comment]
Everything works as we discussed today. `true` in step was causing those problems.

**Comment by Datseris** [comment]
perfect!

File: src/algorithms/optimized_shooting.jl [hunk 8]
```diff
@@ -0,0 +1,98 @@
+export periodic_orbit, periodic_orbits, OptimizedShooting
+
+using LeastSquaresOptim: optimize, LevenbergMarquardt
+
+
+"""
+    OptimizedShooting(; kwargs...)
+
+A shooting method [Dednam2014](@cite) combined with Levenberg-Marquardt optimization 
+to find periodic orbits of continuous systems.
```
Comment (Datseris) [inline src/algorithms/optimized_shooting.jl:10]:
```suggestion
to find periodic orbits of continuous time systems.
```

File: src/algorithms/optimized_shooting.jl [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    end
+    return pos
+end
+
+function costfunc(v, ds, alg)
+    u0 = v[1:dimension(ds)]
+    T = v[end]
+    R1 = compute_residual(ds, u0, T*alg.Δt, alg.n, 0)
+    R2 = compute_residual(ds, u0, T*alg.Δt, alg.n, T)
+    err = R2 .- R1
```
Comment (Datseris) [inline src/algorithms/optimized_shooting.jl:83]: Hm, isn't this here too ineficient? Every time you call the cost function you are creating `R1, R2, err`. Does the `optimize` function run in parallel? If not, we can re-use existing vectors `R1, R2, err` instead of creating them all the time. For example, you can make `costfunc` be a type and use [function-like objects](https://docs.julialang.org/en/v1/manual/methods/#Function-like-objects). Something like

```
struct CostFunc
R1::Vector
R2::Vector
err:Vector
end

function (cf::CostFunc)(v, ds, alg)
   (; R1, R2, err) = cf
   # same code as in your function
end
```

File: src/algorithms/optimized_shooting.jl [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    return err
+end
+
+function compute_residual(ds, u0, Δt, n, t0)
```
Comment (Datseris) [inline src/algorithms/optimized_shooting.jl:87]: Same story. You need to make this `compute_residual!(R, ...)` that uses an existing array `R` instead of creating a new one every time.

File: src/api.jl [hunk 11]
```diff
@@ -20,7 +20,7 @@ A structure that contains an initial guess for a periodic orbit detection algori
     * `u0` - guess of a point in the periodic orbit
     * `T` - guess of period of the orbit
 """
-struct InitialGuess{U<:AbstractArray{<:Real}, R<:Union{Real, Nothing}}
+struct InitialGuess{U<:AbstractArray{<:Real},R<:Union{Real,Nothing}}
```
Comment (Datseris) [inline src/api.jl:23]: In this file you removed spaces between types. Why? Keep them, it is more readable with spaces.

**Review by Datseris — COMMENTED** [review]
You can optimize the code. If you don't do it until our next meeting, then I will show you how to use profiling, and we can see the big difference in profiling before and after the optimizations.

**Comment by JonasKoziorek** [comment]
I tried to make it faster but with no success. Even though I reduced allocations almost in half, the runtime wasn't faster. Let's try profiling together next time then!

**Comment by JonasKoziorek** [comment]
The `minimal_period` was messing up one of the tests so I disabled it for now. I will allow passing kwargs to `minimal_period` in a different PR to be able to affect the precision.

**Comment by JonasKoziorek** [comment]
I suddenly can't build the docs. I attach the [stacktrace.txt](https://github.com/user-attachments/files/16992776/stacktrace.txt) here. It originates in `build_docs_with_style.jl` line 30. Commenting this line solves the problem. Could you help me with this @datseris?

**Comment by Datseris** [comment]
docbuild problem comes from Documenter.jl. We can't resolve it :( :( :( cc @Tortar , I searched throught the JuliaDynamics/doctheme repo, we don't have `$todo` anywhere. It has to be from Documenter.jl.
